iceoryx_dds_gateway/source/iox2dds_app/main.cpp

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <atomic>
+
+#include <iceoryx_posh/runtime/posh_runtime.hpp>
+#include <iceoryx_utils/cxx/helplets.hpp>
+#include <iceoryx_utils/cxx/optional.hpp>
+#include <iceoryx_utils/posix_wrapper/semaphore.hpp>
+#include <ioxdds/gateway/iox2dds.hpp>
+
+class ShutdownManager
+{
+  public:
+    static void scheduleShutdown(int num)
+    {
+        char reason;
+        psignal(num, &reason);
+        s_semaphore.post();
+    }
+    static void waitUntilShutdown()
+    {
+        s_semaphore.wait();
+    }
+
+  private:
+    static iox::posix::Semaphore s_semaphore;
+    ShutdownManager() = default;
+};
+iox::posix::Semaphore ShutdownManager::s_semaphore = iox::posix::Semaphore::create(0u).get_value();
+
+int main(int argc, char* argv[])
+{
+    // Set OS signal handlers
+    signal(SIGINT, ShutdownManager::scheduleShutdown);
+    signal(SIGTERM, ShutdownManager::scheduleShutdown);
+
+    // Start application
+    iox::runtime::PoshRuntime::getInstance("/gateway_iox2dds");
+
+    iox::dds::Iceoryx2DDSGateway<> gateway;
+    gateway.runMultithreaded();
```

Comment:
Does the gateway still work when not calling runMultithreaded? If not would it be possible to move it into the constructor? Or what is the argument for calling it separately?

Reply:
The user has the option to run the gateway manually via the `discover` and `forward` methods.